core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,303 @@
+    case E::INVALID_ARGUMENT:
+      return "provided argument is invalid";
+    case E::RECEIVE_OVERFLOW:
+      return "received unconsumed data amount is greater than it can be";
+    case E::IS_WRITING:
+      return "there is already a pending write operation on this stream";
+    case E::IS_READING:
+      return "there is already a pending read operation on this stream";
+    case E::INVALID_WINDOW_SIZE:
+      return "either window size greater than the maximum one or less than "
+             "current number of unconsumed bytes was tried to be set";
+    case E::INTERNAL_ERROR:
+      return "internal error happened";
+  }
+  return "unknown error";
+}
+
+namespace libp2p::connection {
+
+  YamuxStream::YamuxStream(
+      std::shared_ptr<YamuxedConnection> yamuxed_connection,
+      YamuxedConnection::StreamId stream_id, uint32_t maximum_window_size)
+      : yamuxed_connection_{std::move(yamuxed_connection)},
+        stream_id_{stream_id},
+        maximum_window_size_{maximum_window_size} {}
+
+  void YamuxStream::read(gsl::span<uint8_t> out, size_t bytes,
+                         ReadCallbackFunc cb) {
+    if (bytes == 0 || out.empty() || static_cast<size_t>(out.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (!is_readable_) {
+      return cb(Error::NOT_READABLE);
+    }
+    if (is_reading_) {
+      return cb(Error::IS_READING);
+    }
+
+    is_reading_ = true;
+
+    // if there is already enough data in our buffer, read it immediately and
+    // send an acknowledgement
+    if (read_buffer_.size() >= bytes) {
+      if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), bytes),
+                                   read_buffer_.data(), bytes)
+          != bytes) {
+        return cb(Error::INTERNAL_ERROR);
+      }
+      return yamuxed_connection_->streamAckBytes(
+          stream_id_, bytes,
+          [self{shared_from_this()}, cb = std::move(cb), bytes](auto &&res) {
+            self->is_reading_ = false;
+            if (!res) {
+              return cb(res.error());
+            }
+            self->read_buffer_.consume(bytes);
+            cb(bytes);
+          });
+    }
+
+    // else, set a callback, which is called each time a new data arrives
+    yamuxed_connection_->streamAddDataNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, cb = std::move(cb), out, bytes]() mutable {
+          if (self->read_buffer_.size() < bytes) {
+            // not yet
+            return false;
+          }
+
+          if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), bytes),
+                                       self->read_buffer_.data(), bytes)
+              != bytes) {
+            cb(Error::INTERNAL_ERROR);
+          } else {
+            self->yamuxed_connection_->streamAckBytes(
+                self->stream_id_, bytes,
+                [self, cb = std::move(cb), out, bytes](auto &&res) {
+                  (void)out;
```

Comment:
do you really need to capture `out` ?

Reply:
Will check, I think this was left after some debugging